old/launcher/summoner.1.in
.\" $Id: summoner.1.in 6793 2006-09-28 00:35:39Z guillem $
.TH maemo\-summoner 1 2006-09-28 @VERSION@ "Maemo Launcher manual"
.SH NAME
maemo\-summoner \- summon a program from command line
.SH SYNOPSIS
.B maemo\-summoner
.RI [ options "] " file " [" file-options... ]
.SH DESCRIPTION
\fBmaemo\-summoner\fP directly loads an application in the same way that
the \fBmaemo\-launcher\fP would do, making easier the debugging tasks.
It will summon the file provided on the command line with its optional
arguments, otherwise it will summon its own program name with the string
\fI.launch\fP appended and will pass the arguments as well.
So it can be symlinked in place of the real \fIapplication\fP, and this
one moved to \fIapplication.launch\fP.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Show version of program.
.SH SEE ALSO
.BR maemo\-invoker "(1), " maemo\-launcher (1).
.br
.SH AUTHOR
\fBmaemo\-summoner\fP was written by Guillem Jover <guillem.jover@nokia.com>.


old/ChangeLog
2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.32.

2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Add --enable-qt-booster.
	* launcher/Makefile.am (booster_qt_so_SOURCES): New variable.
	(booster_qt_so_CFLAGS): Likewise.
	(booster_qt_so_LDFLAGS): Likewise.
	(booster_qt_so_CPPFLAGS): Likewise.
	(booster_qt_so_LDADD): Likewise.
	[USE_QT_BOOSTER] (booster_PROGRAMS): Likewise.
	* launcher/booster_qt.c: New file.

2008-12-19  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.32~.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.31.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/invoker.c (disable_self_core_dump): New function.
	(main): Call disable_self_core_dump before setting up the signal
	forwarder.

2008-07-03  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.31~.

2008-05-29  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.30.

2008-05-28  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac (HILDON_DEPENDS): Replace libossohelp with hildon-help.

2008-05-12  Eero Tamminen  <eero.tamminen@nokia.com>

	* README: Typo fix and additional information for symbol versioning.

2008-05-12  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.30~.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.29.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (kindergarten_destroy): Traverse the whole
	child list and act on childs with a non-zero pid.
	(store_state): Likewise.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (child_destroy): Reset pid at the beginning.

2008-05-05  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.29~.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.28.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (rise_oom_defense): Change error to warning
	when the defender failed to execute.

2008-04-27  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/report.h (enum report_type): Add report_warning.
	(warning): New macro.
	* launcher/report.c (vreport): Handle report_warning.

2008-04-23  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/launcher.c (store_state): Call comm_msg_destroy.
	(load_state): Likewise.

2008-04-23  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Bump version to 0.28~.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac: Version 0.27.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* README: Add some documentation about version scripts.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* README: Add some documentation for make based build systems.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_set_magic): Rename to ...
	(comm_msg_put_magic): ... this. Fix all users.
	(comm_msg_pack_int): Rename to ...
	(comm_msg_put_int): ... this. Fix all users.
	(comm_msg_unpack_int): Rename to ...
	(comm_msg_get_int): ... this. Fix all users.
	(comm_msg_pack_str): Rename to ...
	(comm_msg_put_str): ... this. Fix all users.
	(comm_msg_unpack_str): Rename to ...
	(comm_msg_get_str): ... this. Fix all users.

2008-04-15  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_unpack_mem): Remove function.
	(comm_msg_put_u32): New function.
	(comm_msg_get_u32): Likewise.
	(comm_msg_set_magic): Use comm_msg_put_u32 instead of direct buffer
	access.
	(comm_msg_pack_int): Likewise.
	(comm_msg_pack_str): Likewise.
	(comm_msg_get_magic): Use comm_msg_get_u32 instead of direct buffer
	access.
	(comm_msg_unpack_int): Likewise. Perform boundary checks.
	(comm_msg_unpack_str): Likewise. Get the string through the current
	buffer read address.

2008-03-20  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_set_magic): New prototype.
	(comm_msg_get_magic): Likewise.
	* launcher/comm_msg.c (comm_msg_set_magic): New function.
	(comm_msg_get_magic): Likewise.
	* launcher/test_msg.c (main): Set and get the magic and check the
	value being the same.
	* launcher/launcher.c (LAUNCHER_STATE_SIG): Change from "MLSF0.0"
	to 0x30534c4d.
	(store_state): Use comm_msg_set_magic instead of comm_msg_pack_str.
	(load_state): Use comm_msg_get_magic instead of comm_msg_unpack_str.

2008-03-18  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.h (comm_msg_print): New prototype.
	* launcher/comm_msg.c (struct comm_msg): Change buf member from char *
	to void *.
	(comm_msg_print): New function.
	(comm_msg_send): Use comm_msg_print to print msg instead of debug.
	(comm_msg_recv): Likewise.

2008-03-14  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_send): Use msg->used in sizeof instead
	of msg->size.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* README: Fix a typo and add a missing word.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* client/client.c (cleanup): New function.
	(main): Install cleanup as an atexit function.
	* launcher/launcher.c (launch_process): Call exit instead of _exit.
	* launcher/summoner.c (summon_process): Likewise.
	(main): Return 0 at the end of the function instead of _exit.

2008-03-07  Guillem Jover  <guillem.jover@nokia.com>

	* configure.ac (CLIENT): Use only the minimum dependencies,
	GTK_DEPENDS and hildon-1.

2008-02-22  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/comm_msg.c (comm_msg_pack_mem): Move code to ...
	(comm_msg_pack_str): ... here.
	(comm_msg_pack_int): Rewrite specialized version without taking into
	account unaligned memory to improve performance.

2008-02-22  Guillem Jover  <guillem.jover@nokia.com>

	* launcher/Makefile.am (check_PROGRAMS): Add 'test-msg-perf'.
	(test_msg_perf_SOURCES): New variable.
	(test_msg_perf_CPPFLAGS): Likewise.

old/launcher/summoner.c
/*
 * Copyright © 2006, 2007 Nokia Corporation
 *
 * Authors: Guillem Jover <guillem.jover@nokia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 *
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "config.h"
#include "report.h"
#include "prog.h"
#include "search.h"

static void
summon_process(prog_t *prog)
{
  load_main(prog);

  print_prog_env_argv(prog);

  debug("summoning process: '%s'\n", prog->filename);

  exit(prog->entry(prog->argc, prog->argv));
}

static void
version(void)
{
  printf("%s (%s) %s\n", PROG_NAME, PACKAGE, VERSION);

  exit(0);
}

static void
usage(int status)
{
  printf("Usage: %s [options] file [file-options]\n"
	 "\n"
	 "Options:\n"
	 "  --version           Print program version.\n"
	 "  --help              Print this help message.\n",
	 PROG_NAME);

  exit(status);
}

int
main(int argc, char *argv[])
{
  int i;
  prog_t prog = { 0 };

  if (strstr(argv[0], PROG_NAME))
  {
    char *argv0 = NULL;

    /* Called with our default name. Parse arguments. */
    for (i = 1; i < argc; ++i)
    {
      if (strcmp(argv[i], "--version") == 0)
	version();
      else if (strcmp(argv[i], "--help") == 0)
	usage(0);
      else if (strncmp(argv[i], "--", 2) == 0)
	usage(1);
      else
      {
	char *period;

	argv0 = strdup(argv[i]);
	period = strstr(argv0, ".launch");
	if (period)
		*period = '\0';

	prog.filename = search_program(argv[i]);
	prog.argc = argc - i;
	prog.argv = argv;
	break;
      }
    }

    if (!prog.filename)
      usage(1);

    set_progname(argv0, argc, argv, i + 1);

    free(argv0);
  }
  else
  {
    char *launch = NULL;

    /* Called with a different name. Add the proper extension and launch it.
     * Do not try to parse any arguments. */
    if (asprintf(&launch, "%s.launch", argv[0]) < 0)
      die(1, "allocating program name buffer");
    prog.filename = search_program(launch);
    prog.argc = argc;
    prog.argv = argv;

    free(launch);
  }

  /* Summon it. */
  if (prog.filename)
  {
    info("summoning '%s'\n", prog.filename);
    summon_process(&prog);
  }
  else
    error("nothing to summon\n");

  return 0;
}

